File: src/electron/instances.ts
// Game Instance Management - each instance is a separate game profile
// with its own mods, config, and saves

import { app } from "electron";
import path from "node:path";
import fs from "node:fs";
import { createRequire } from "node:module";
import { getMinecraftDir } from "./config.js";

// Native module loading - Rust core for performance
const customRequire = createRequire(__filename);
let nativeModule: any = null;
const DEBUG_INSTANCES = process.env.DEBUG_INSTANCES === "1";

function getNative(): any {
  if (nativeModule) return nativeModule;
  const nativePath = path.join(app.getAppPath(), "native", "index.cjs");
  nativeModule = customRequire(nativePath);
  return nativeModule;
}

// Type definitions
export type LoaderType = "vanilla" | "fabric" | "forge" | "neoforge" | "quilt";

export interface GameInstance {
  id: string;
  name: string;
  icon?: string;
  minecraftVersion: string;
  loader: LoaderType;
  loaderVersion?: string;
  createdAt: string;
  lastPlayedAt?: string;
  totalPlayTime: number;
  javaPath?: string;
  ramMB?: number;
  javaArguments?: string;
  gameDirectory: string;
  modpackId?: string;
  modpackVersionId?: string;
  cloudId?: string;
  autoUpdate?: boolean;
  banner?: string;
  lockedMods?: string[];
}

export interface CreateInstanceOptions {
  name: string;
  minecraftVersion: string;
  loader?: LoaderType;
  loaderVersion?: string;
  icon?: string;
  javaPath?: string;
  ramMB?: number;
  modpackId?: string;
  modpackVersionId?: string;
}

export interface UpdateInstanceOptions {
  name?: string;
  icon?: string;
  loader?: LoaderType;
  loaderVersion?: string;
  javaPath?: string;
  ramMB?: number;
  javaArguments?: string;
  lastPlayedAt?: string;
  totalPlayTime?: number;
  autoUpdate?: boolean;
  lockedMods?: string[];
}

// Global cache for all loaded instances - ensures reliable lookup
// even when current view is a paginated subset
const instanceCache = new Map<string, GameInstance>();

// Current list of instances (may be paginated subset for UI)
let instances: GameInstance[] = [];

// Path helpers
export function getInstancesDir(): string {
  const minecraftDir = getMinecraftDir();
  return path.join(minecraftDir, "instances");
}

export function getInstanceDir(instanceId: string): string {
  return path.join(getInstancesDir(), instanceId);
}

// Deprecated: Use instance.json in each instance folder instead
function getInstancesMetaPath(): string {
  return path.join(getInstancesDir(), "instances.json");
}

function getInstanceMetaPath(instanceId: string): string {
  return path.join(getInstanceDir(instanceId), "instance.json");
}

export function getInstanceIconPath(instanceId: string): string {
  return path.join(getInstanceDir(instanceId), "icon.png");
}

// Load instances from disk - uses Native Rust Core for performance
export async function loadInstances(
  offset: number = 0,
  limit: number = 1000,
): Promise<GameInstance[]> {
  let nativeInstances: any[] = [];
  try {
    const instancesDir = getInstancesDir();

    // Create instances directory if not exists
    try {
      await fs.promises.access(instancesDir);
    } catch {
      await fs.promises.mkdir(instancesDir, { recursive: true });
      instances = [];
      return instances;
    }

    // Migrate from old instances.json if exists
    const oldMetaPath = getInstancesMetaPath();
    let oldMetaExists = false;
    try {
      await fs.promises.access(oldMetaPath);
      oldMetaExists = true;
    } catch {}
    if (oldMetaExists) {
      console.log("[Instances] Migrating from instances.json to per-instance files...");
      try {
        const oldData = await fs.promises.readFile(oldMetaPath, "utf-8");
        const oldInstances = JSON.parse(oldData) as GameInstance[];

        for (const inst of oldInstances) {
          const instDir = getInstanceDir(inst.id);
          await fs.promises.mkdir(instDir, { recursive: true });
          const metaPath = getInstanceMetaPath(inst.id);
          await fs.promises.writeFile(metaPath, JSON.stringify(inst, null, 2));
          console.log("[Instances] Migrated:", inst.name);
        }

        await fs.promises.unlink(oldMetaPath);
        console.log("[Instances] Migration complete, deleted instances.json");
      } catch (migError) {
        console.error("[Instances] Migration error:", migError);
      }
    }

    // Use Native Rust Core to list instances efficiently
    const native = getNative();
    nativeInstances = (await native.listInstances(
      instancesDir,
      offset,
      limit,
    )) as any[];

    // Convert Native meta to GameInstance
    const loadedInstances: GameInstance[] = nativeInstances.map((meta) => ({
      id: meta.id,
      name: meta.name,
      icon: meta.icon,
      banner: meta.banner,
      minecraftVersion: meta.minecraftVersion,
      loader: meta.loader.toLowerCase() as LoaderType,
      loaderVersion: meta.loaderVersion,
      createdAt: meta.createdAt,
      lastPlayedAt: meta.lastPlayedAt,
      totalPlayTime: meta.totalPlayTime,
      javaPath: meta.javaPath,
      ramMB: meta.ramMB,
      javaArguments: meta.javaArguments,
      gameDirectory: meta.gameDirectory,
      modpackId: meta.modpackId,
      modpackVersionId: meta.modpackVersionId,
      cloudId: meta.cloudId,
      autoUpdate: meta.autoUpdate,
      lockedMods: meta.lockedMods,
    }));

    // Update cache and current view
    for (const inst of loadedInstances) {
      instanceCache.set(inst.id, inst);
    }
    instances = loadedInstances;

    if (DEBUG_INSTANCES) {
      console.log(
        `[Instances] Loaded ${instances.length} instances offset=${offset} limit=${limit}. Cache size: ${instanceCache.size}`,
      );
    }
  } catch (error) {
    console.error("[Instances] Failed to load instances:", error);
    instances = [];
  }

  if (DEBUG_INSTANCES && instances.length === 0) {
    console.error(
      "[Debug] Loaded 0 instances! Native returned:",
      JSON.stringify(nativeInstances),
    );
  }

  return instances;
}

// Cache to avoid redundant disk writes
const lastSavedContent = new Map<string, string>();

// Save single instance to its instance.json file
async function saveInstance(instance: GameInstance): Promise<void> {
  try {
    const instanceDir = getInstanceDir(instance.id);
    await fs.promises.mkdir(instanceDir, { recursive: true });

    const metaPath = getInstanceMetaPath(instance.id);

    // Don't save icon if loaded from file (stored as icon.png separately)
    const saveData = { ...instance };
    if (
      saveData.icon?.startsWith("file://") ||
      saveData.icon?.startsWith("data:")
    ) {
      delete saveData.icon;
    }

    const json = JSON.stringify(saveData, null, 2);

    // Only write if content has changed
    if (lastSavedContent.get(instance.id) === json) {
      return;
    }

    // Atomic write: write to .tmp then rename
    const tmpPath = `${metaPath}.tmp`;
    await fs.promises.writeFile(tmpPath, json);
    await fs.promises.rename(tmpPath, metaPath);

    lastSavedContent.set(instance.id, json);
    if (DEBUG_INSTANCES) {
      console.log("[Instances] Saved instance:", instance.name);
    }
  } catch (error) {
    console.error("[Instances] Failed to save instance:", instance.id, error);
  }
}

async function saveInstances(): Promise<void> {
  await Promise.all(instances.map((inst) => saveInstance(inst)));
  if (DEBUG_INSTANCES) {
    console.log("[Instances] Saved all", instances.length, "instances");
  }
}

// CRUD operations
export async function getInstances(
  offset: number = 0,
  limit: number = 1000,
): Promise<GameInstance[]> {
  return await loadInstances(offset, limit);
}

export function getInstance(id: string): GameInstance | null {
  if (instanceCache.has(id)) {
    return instanceCache.get(id)!;
  }

  const instance = instances.find((i) => i.id === id);
  if (!instance) {
    console.warn(
      `[Instances] getInstance FAILED for ID: "${id}". Cache Size: ${instanceCache.size}. Available: ${instances.length}`,
    );
  }
  return instance || null;
}

export async function createInstance(
  options: CreateInstanceOptions,
): Promise<GameInstance> {
  const id = generateUniqueId(options.name);
  const gameDirectory = getInstanceDir(id);

  await fs.promises.mkdir(gameDirectory, { recursive: true });

  // Create subdirectories in parallel
  const subdirs = ["mods", "config", "saves", "resourcepacks", "shaderpacks", "datapacks"];
  await Promise.all(
    subdirs.map((subdir) =>
      fs.promises.mkdir(path.join(gameDirectory, subdir), { recursive: true }),
    ),
  );

  const instance: GameInstance = {
    id,
    name: options.name,
    icon: options.icon,
    minecraftVersion: options.minecraftVersion,
    loader: options.loader || "vanilla",
    loaderVersion: options.loaderVersion,
    createdAt: new Date().toISOString(),
    totalPlayTime: 0,
    javaPath: options.javaPath,
    ramMB: options.ramMB,
    gameDirectory,
    modpackId: options.modpackId,
    modpackVersionId: options.modpackVersionId,
  };

  instances.push(instance);
  instanceCache.set(instance.id, instance);
  await saveInstance(instance);

  if (DEBUG_INSTANCES) {
    console.log("[Instances] Created instance:", instance.name, instance.id);
  }
  return instance;
}

export async function importCloudInstance(
  cloudData: any,
): Promise<GameInstance> {
  const id = cloudData.storagePath || cloudData.id;
  const gameDirectory = getInstanceDir(id);
  const instancesDir = getInstancesDir();

  const native = getNative();
  if (typeof native.createInstanceDirectories !== "function") {
    console.error(
      "[Instances] CRITICAL: createInstanceDirectories is NOT a function. Available:",
      Object.keys(native),
    );
  }
  native.createInstanceDirectories(instancesDir, id);

  // Search by cloudId first, then fallback to id
  let existingIndex = instances.findIndex((i) => i.cloudId === cloudData.id);
  if (existingIndex === -1) {
    existingIndex = instances.findIndex((i) => i.id === id);
  }

  const instance: GameInstance = {
    id,
    name: cloudData.name,
    icon: cloudData.iconUrl || undefined,
    minecraftVersion: cloudData.minecraftVersion || "1.20.1",
    loader: (cloudData.loaderType as LoaderType) || "fabric",
    loaderVersion: cloudData.loaderVersion,
    createdAt: cloudData.createdAt || new Date().toISOString(),
    totalPlayTime: 0,
    gameDirectory,
    modpackId:
      cloudData.modpackType === "modrinth" ? cloudData.modpackUrl : undefined,
    cloudId: cloudData.id,
    autoUpdate:
      existingIndex !== -1 ? instances[existingIndex].autoUpdate : true,
    banner: cloudData.bannerUrl || cloudData.image,
  };

  if (existingIndex !== -1) {
    // Merge with existing - preserve local fields
    instances[existingIndex] = {
      ...instances[existingIndex],
      ...instance,
      id: instances[existingIndex].id,
      gameDirectory: instances[existingIndex].gameDirectory,
      autoUpdate:
        instances[existingIndex].autoUpdate !== undefined
          ? instances[existingIndex].autoUpdate
          : instance.autoUpdate,
    };
  } else {
    instances.push(instance);
  }

  const finalInstance =
    existingIndex !== -1 ? instances[existingIndex] : instance;

  instanceCache.set(finalInstance.id, finalInstance);
  await saveInstance(finalInstance);

  if (DEBUG_INSTANCES) {
    console.log(
      "[Instances] Imported cloud instance:",
      finalInstance.name,
      finalInstance.id,
    );
  }
  return finalInstance;
}

export async function updateInstance(
  id: string,
  updates: UpdateInstanceOptions,
): Promise<GameInstance | null> {
  const index = instances.findIndex((i) => i.id === id);
  let instance = index !== -1 ? instances[index] : instanceCache.get(id);

  if (!instance) {
    console.warn(
      `[Instances] updateInstance: Instance ${id} not found in array or cache.`,
    );
    return null;
  }

  const updatedInstance = { ...instance, ...updates };
  instanceCache.set(id, updatedInstance);

  if (index !== -1) {
    instances[index] = updatedInstance;
  }

  await saveInstance(updatedInstance);

  if (DEBUG_INSTANCES) {
    console.log("[Instances] Updated instance:", id);
  }
  return updatedInstance;
}

export async function deleteInstance(id: string): Promise<boolean> {
  const index = instances.findIndex((i) => i.id === id);
  const instance = index !== -1 ? instances[index] : instanceCache.get(id);

  if (!instance) {
    console.warn(`[Instances] deleteInstance: Instance ${id} not found.`);
    return false;
  }

  const targetDir = getInstanceDir(id);
  console.log("[Instances] Deleting:", instance.name, "at", targetDir);

  try {
    let dirExists = false;
    try {
      await fs.promises.access(targetDir);
      dirExists = true;
    } catch {}

    if (dirExists) {
      await fs.promises.rm(targetDir, {
        recursive: true,
        force: true,
        maxRetries: 5,
        retryDelay: 200,
      });
      console.log("[Instances] Deleted directory:", targetDir);
    } else {
      console.log("[Instances] Directory already gone:", targetDir);
    }
  } catch (error) {
    console.error("[Instances] Failed to delete directory:", error);
    return false;
  }

  if (index !== -1) {
    instances.splice(index, 1);
  }
  lastSavedContent.delete(id);
  instanceCache.delete(id);

  if (instance.cloudId) {
    console.log(`[Instances] Deleted cloud instance ${instance.cloudId}`);
  }

  console.log("[Instances] Deleted from state:", instance.name, id);
  return true;
}

export async function duplicateInstance(
  id: string,
): Promise<GameInstance | null> {
  const source = getInstance(id);
  if (!source) {
    return null;
  }

  const newInstance = await createInstance({
    name: `${source.name} (Copy)`,
    minecraftVersion: source.minecraftVersion,
    loader: source.loader,
    loaderVersion: source.loaderVersion,
    icon: source.icon,
    javaPath: source.javaPath,
    ramMB: source.ramMB,
  });

  try {
    const sourceDir = source.gameDirectory;
    const destDir = newInstance.gameDirectory;

    const copyDirs = ["mods", "config", "resourcepacks", "shaderpacks", "saves", "datapacks"];
    for (const dir of copyDirs) {
      const srcPath = path.join(sourceDir, dir);
      const dstPath = path.join(destDir, dir);
      try {
        await fs.promises.access(srcPath);
        await copyDirectoryAsync(srcPath, dstPath);
      } catch {
        /* skip if source doesn't exist */
      }
    }

    const copyFiles = ["options.txt", "servers.dat"];
    for (const file of copyFiles) {
      const srcPath = path.join(sourceDir, file);
      const dstPath = path.join(destDir, file);
      try {
        await fs.promises.copyFile(srcPath, dstPath);
      } catch {
        /* skip if file doesn't exist */
      }
    }
  } catch (error) {
    console.error("[Instances] Failed to copy files:", error);
  }

  console.log("[Instances] Duplicated:", source.name, "->", newInstance.name);
  return newInstance;
}

export async function markInstancePlayed(
  id: string,
  playTimeMinutes: number = 0,
): Promise<void> {
  const instance = getInstance(id);
  if (instance) {
    await updateInstance(id, {
      lastPlayedAt: new Date().toISOString(),
      totalPlayTime: instance.totalPlayTime + playTimeMinutes,
    });
  }
}

export async function setInstanceIcon(
  id: string,
  iconData: string,
): Promise<{ ok: boolean; error?: string }> {
  const instance = getInstance(id);
  if (!instance) {
    return { ok: false, error: "Instance not found" };
  }

  const iconPath = getInstanceIconPath(id);

  try {
    let imageBuffer: Buffer;

    if (iconData.startsWith("data:image")) {
      const base64Data = iconData.replace(/^data:image\/\w+;base64,/, "");
      imageBuffer = Buffer.from(base64Data, "base64");
    } else {
      try {
        await fs.promises.access(iconData);
      } catch {
        return { ok: false, error: "Invalid icon data" };
      }

      const resolved = path.resolve(iconData);

      // Security: only allow files from safe directories
      const allowedDirs = [
        app.getPath("pictures"),
        app.getPath("downloads"),
        app.getPath("desktop"),
        app.getPath("home"),
        getMinecraftDir(),
      ];

      const isSafe = allowedDirs.some((dir) => {
        const allowedPath = path.resolve(dir);
        return resolved.startsWith(allowedPath);
      });

      if (!isSafe) {
        console.warn("[Instances] Blocked path traversal attempt:", iconData);
        return { ok: false, error: "Invalid file path - access denied" };
      }

      imageBuffer = await fs.promises.readFile(iconData);
    }

    await fs.promises.writeFile(iconPath, imageBuffer);
    await loadInstances();

    console.log("[Instances] Set icon for:", instance.name);
    return { ok: true };
  } catch (error) {
    console.error("[Instances] Failed to set icon:", error);
    return { ok: false, error: String(error) };
  }
}

// Helper functions
async function copyDirectoryAsync(src: string, dest: string): Promise<void> {
  await fs.promises.mkdir(dest, { recursive: true });

  const entries = await fs.promises.readdir(src, { withFileTypes: true });
  for (const entry of entries) {
    const srcPath = path.join(src, entry.name);
    const destPath = path.join(dest, entry.name);

    if (entry.isDirectory()) {
      await copyDirectoryAsync(srcPath, destPath);
    } else {
      await fs.promises.copyFile(srcPath, destPath);
    }
  }
}

function sanitizeName(name: string): string {
  return name
    .trim()
    .replace(/[<>:"/\\|?*]/g, "")
    .replace(/\s+/g, " ")
    .substring(0, 100);
}

function generateUniqueId(name: string): string {
  const baseName = sanitizeName(name);

  if (!instances.find((i) => i.id === baseName)) {
    return baseName;
  }

  let counter = 1;
  let uniqueName = `${baseName}-${counter}`;
  while (instances.find((i) => i.id === uniqueName)) {
    counter++;
    uniqueName = `${baseName}-${counter}`;
  }

  return uniqueName;
}

// Initialize on module load
loadInstances().catch((err) =>
  console.error("[Instances] Init load failed:", err),
);

File: src-tauri/src/instances.rs



use crate::config::get_minecraft_dir;
use base64::Engine;
use once_cell::sync::Lazy;
use serde::{Deserialize, Serialize};
use std::fs;
use std::path::PathBuf;
use std::sync::Mutex;


#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "lowercase")]
pub enum LoaderType {
    Vanilla,
    Fabric,
    Forge,
    Neoforge,
    Quilt,
}

impl Default for LoaderType {
    fn default() -> Self {
        Self::Vanilla
    }
}


#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GameInstance {
    pub id: String,
    pub name: String,
    pub icon: Option<String>,
    pub minecraft_version: String,
    pub loader: LoaderType,
    pub loader_version: Option<String>,
    pub created_at: String,
    pub last_played_at: Option<String>,
    pub total_play_time: u32, 
    pub java_path: Option<String>,
    #[serde(rename = "ramMB")]
    pub ram_mb: Option<u32>,
    pub java_arguments: Option<String>,
    pub game_directory: String,
    pub modpack_id: Option<String>,
    pub modpack_version_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub cloud_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub auto_update: Option<bool>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub banner: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub locked_mods: Option<Vec<String>>,
}


#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CreateInstanceOptions {
    pub name: String,
    pub minecraft_version: String,
    pub loader: Option<LoaderType>,
    pub loader_version: Option<String>,
    pub icon: Option<String>,
}


pub(crate) static INSTANCES: Lazy<Mutex<Vec<GameInstance>>> =
    Lazy::new(|| Mutex::new(load_instances_from_disk()));


pub fn get_instances_dir() -> PathBuf {
    get_minecraft_dir().join("instances")
}


pub fn get_instance_dir(id: &str) -> PathBuf {
    get_instances_dir().join(id)
}



fn detect_image_mime(bytes: &[u8]) -> &'static str {
    if bytes.len() >= 12 && &bytes[0..4] == b"RIFF" && &bytes[8..12] == b"WEBP" {
        "image/webp"
    } else if bytes.starts_with(&[0xFF, 0xD8, 0xFF]) {
        "image/jpeg"
    } else if bytes.starts_with(&[0x47, 0x49, 0x46]) {
        "image/gif"
    } else {
        "image/png"
    }
}

fn load_instances_from_disk() -> Vec<GameInstance> {
    let instances_dir = get_instances_dir();

    if !instances_dir.exists() {
        if let Err(e) = fs::create_dir_all(&instances_dir) {
            eprintln!("[Instances] Failed to create instances dir: {}", e);
        }
        return Vec::new();
    }

    let mut instances = Vec::new();

    let Ok(entries) = fs::read_dir(&instances_dir) else {
        return instances;
    };
    for entry in entries.flatten() {
        if !entry.path().is_dir() {
            continue;
        }
        let meta_path = entry.path().join("instance.json");
        if !meta_path.exists() {
            continue;
        }
        let Ok(content) = fs::read_to_string(&meta_path) else {
            continue;
        };
        let Ok(mut instance) = serde_json::from_str::<GameInstance>(&content) else {
            continue;
        };
        instance.game_directory = entry.path().to_string_lossy().to_string();
        instance.id = entry.file_name().to_string_lossy().to_string();

        let icon_path = entry.path().join("icon.png");
        if icon_path.exists() && instance.icon.as_ref().map_or(true, |i| i.is_empty()) {
            if let Ok(icon_data) = fs::read(&icon_path) {
                let mime = detect_image_mime(&icon_data);
                let base64 = base64::engine::general_purpose::STANDARD.encode(&icon_data);
                instance.icon = Some(format!("data:{};base64,{}", mime, base64));
            }
        }

        instances.push(instance);
    }

    println!("[Instances] Loaded {} instances", instances.len());
    instances
}

fn save_instance(instance: &GameInstance) -> Result<(), String> {
    let instance_dir = get_instance_dir(&instance.id);
    fs::create_dir_all(&instance_dir).map_err(|e| e.to_string())?;

  
    let mut save_data = instance.clone();
    if let Some(ref icon) = save_data.icon {
        if icon.starts_with("data:") || icon.starts_with("file://") {
            save_data.icon = None;
        }
    }

    let meta_path = instance_dir.join("instance.json");
    let content = serde_json::to_string_pretty(&save_data).map_err(|e| e.to_string())?;
    fs::write(&meta_path, content).map_err(|e| e.to_string())?;

    Ok(())
}


fn sanitize_name(name: &str) -> String {
    name.chars()
        .filter(|c| !['<', '>', ':', '"', '/', '\\', '|', '?', '*'].contains(c))
        .collect::<String>()
        .trim()
        .chars()
        .take(100)
        .collect()
}


fn generate_unique_id(name: &str, existing: &[GameInstance]) -> String {
    let base = sanitize_name(name);

    if !existing.iter().any(|i| i.id == base) {
        return base;
    }

    let mut counter = 1;
    loop {
        let candidate = format!("{}-{}", base, counter);
        if !existing.iter().any(|i| i.id == candidate) {
            return candidate;
        }
        counter += 1;
    }
}


pub fn instances_list_sync() -> Vec<GameInstance> {
    let mut instances = INSTANCES.lock().unwrap();
    *instances = load_instances_from_disk();
    instances.clone()
}

#[tauri::command]
pub async fn instances_list() -> Vec<GameInstance> {
    tokio::task::spawn_blocking(|| {
        instances_list_sync()
    })
    .await
    .unwrap_or_default()
}

#[tauri::command]
pub fn instances_get(id: String) -> Option<GameInstance> {
    let instances = INSTANCES.lock().unwrap();
    instances.iter().find(|i| i.id == id).cloned()
}


pub fn mark_played(id: &str) {
    if let Ok(mut instances) = INSTANCES.lock() {
        if let Some(instance) = instances.iter_mut().find(|i| i.id == id) {
            instance.last_played_at = Some(chrono::Utc::now().to_rfc3339());
            let _ = save_instance(instance);
        }
    }
}

#[tauri::command]
pub fn instances_create(options: CreateInstanceOptions) -> Result<GameInstance, String> {
    let mut instances = INSTANCES.lock().map_err(|e| e.to_string())?;

    let id = generate_unique_id(&options.name, &instances);
    let game_directory = get_instance_dir(&id).to_string_lossy().to_string();

    let instance = GameInstance {
        id: id.clone(),
        name: options.name,
        icon: options.icon,
        minecraft_version: options.minecraft_version,
        loader: options.loader.unwrap_or_default(),
        loader_version: options.loader_version,
        created_at: chrono::Utc::now().to_rfc3339(),
        last_played_at: None,
        total_play_time: 0,
        java_path: None,
        ram_mb: None,
        java_arguments: None,
        game_directory,
        modpack_id: None,
        modpack_version_id: None,
        cloud_id: None,
        auto_update: None,
        banner: None,
        locked_mods: None,
    };

    let instance_dir = get_instance_dir(&id);
    for subdir in ["mods", "config", "saves", "resourcepacks", "shaderpacks"] {
        fs::create_dir_all(instance_dir.join(subdir)).map_err(|e| e.to_string())?;
    }

    save_instance(&instance)?;
    instances.push(instance.clone());

    println!("[Instances] Created: {} ({})", instance.name, instance.id);
    Ok(instance)
}

#[tauri::command]
pub fn instances_update(
    id: String,
    updates: serde_json::Value,
) -> Result<Option<GameInstance>, String> {
    let mut instances = INSTANCES.lock().map_err(|e| e.to_string())?;

    if let Some(instance) = instances.iter_mut().find(|i| i.id == id) {
        if let Some(name) = updates.get("name").and_then(|v| v.as_str()) {
            instance.name = name.to_string();
        }
        if let Some(loader) = updates.get("loader").and_then(|v| v.as_str()) {
            instance.loader = match loader {
                "vanilla" => LoaderType::Vanilla,
                "fabric" => LoaderType::Fabric,
                "forge" => LoaderType::Forge,
                "neoforge" => LoaderType::Neoforge,
                "quilt" => LoaderType::Quilt,
                _ => instance.loader.clone(),
            };
        }
        if let Some(version) = updates.get("minecraftVersion").and_then(|v| v.as_str()) {
            instance.minecraft_version = version.to_string();
        }
        if let Some(cloud_id) = updates.get("cloudId").and_then(|v| v.as_str()) {
            instance.cloud_id = Some(cloud_id.to_string());
        }
        if let Some(banner) = updates.get("banner") {
            if banner.is_null() {
                instance.banner = None;
            } else if let Some(s) = banner.as_str() {
                instance.banner = Some(s.to_string());
            }
        }
        if let Some(icon) = updates.get("icon").and_then(|v| v.as_str()) {
            instance.icon = Some(icon.to_string());
        }
        if let Some(auto_update) = updates.get("autoUpdate").and_then(|v| v.as_bool()) {
            instance.auto_update = Some(auto_update);
        }
        if let Some(mods) = updates.get("lockedMods").and_then(|v| v.as_array()) {
            instance.locked_mods = Some(
                mods.iter()
                    .filter_map(|v| v.as_str().map(|s| s.to_string()))
                    .collect(),
            );
        }
        if let Some(ver) = updates.get("loaderVersion").and_then(|v| v.as_str()) {
            instance.loader_version = Some(ver.to_string());
        }
        if let Some(path) = updates.get("javaPath").and_then(|v| v.as_str()) {
            instance.java_path = Some(path.to_string());
        }
        if let Some(args) = updates.get("javaArguments").and_then(|v| v.as_str()) {
            instance.java_arguments = Some(args.to_string());
        }
        if let Some(mb) = updates.get("ramMB").and_then(|v| v.as_u64()) {
            instance.ram_mb = Some(mb as u32);
        }

        save_instance(instance)?;
        return Ok(Some(instance.clone()));
    }

    Ok(None)
}

#[tauri::command]
pub async fn instances_delete(id: String) -> Result<bool, String> {
    let existed = {
        let mut instances = INSTANCES.lock().map_err(|e| e.to_string())?;
        match instances.iter().position(|i| i.id == id) {
            Some(pos) => {
                let instance = instances.remove(pos);
                println!("[Instances] Deleted: {} ({})", instance.name, id);
                true
            }
            None => false,
        }
    };

    if !existed {
        return Ok(false);
    }

    let dir = get_instance_dir(&id);
    if dir.exists() {
        tauri::async_runtime::spawn_blocking(move || fs::remove_dir_all(&dir))
            .await
            .map_err(|e| e.to_string())?
            .map_err(|e| e.to_string())?;
    }

    Ok(true)
}

#[tauri::command]
pub fn instances_duplicate(id: String) -> Result<Option<GameInstance>, String> {
    let instances = INSTANCES.lock().map_err(|e| e.to_string())?;

    let source = instances.iter().find(|i| i.id == id).cloned();
    drop(instances);

    if let Some(source) = source {
        let new_instance = instances_create(CreateInstanceOptions {
            name: format!("{} (Copy)", source.name),
            minecraft_version: source.minecraft_version,
            loader: Some(source.loader),
            loader_version: source.loader_version,
            icon: source.icon,
        })?;

        let source_dir = get_instance_dir(&id);
        let dest_dir = get_instance_dir(&new_instance.id);

        for dir in ["mods", "config", "resourcepacks", "shaderpacks", "saves"] {
            let src = source_dir.join(dir);
            let dst = dest_dir.join(dir);
            if src.exists() {
                copy_dir_recursive(&src, &dst).ok();
            }
        }

        for file in ["options.txt", "servers.dat"] {
            let src = source_dir.join(file);
            let dst = dest_dir.join(file);
            if src.exists() {
                fs::copy(&src, &dst).ok();
            }
        }

        return Ok(Some(new_instance));
    }

    Ok(None)
}

#[tauri::command]
pub fn instances_set_icon(id: String, icon_data: String) -> crate::cloud::SimpleResult {
    let icon_path = get_instance_dir(&id).join("icon.png");

    let result: Result<(), String> = (|| {
        if icon_data.starts_with("data:image") {
            let base64_data = icon_data.split(',').nth(1).ok_or("Invalid icon data")?;
            let bytes = base64::engine::general_purpose::STANDARD
                .decode(base64_data)
                .map_err(|e| e.to_string())?;
            fs::write(&icon_path, bytes).map_err(|e| e.to_string())?;
        }
        Ok(())
    })();

    match result {
        Ok(()) => crate::cloud::SimpleResult {
            ok: true,
            error: None,
            message: None,
        },
        Err(e) => crate::cloud::SimpleResult {
            ok: false,
            error: Some(e),
            message: None,
        },
    }
}

/// Copy directory recursively
pub(crate) fn copy_dir_recursive(src: &PathBuf, dst: &PathBuf) -> Result<(), std::io::Error> {
    fs::create_dir_all(dst)?;

    for entry in fs::read_dir(src)? {
        let entry = entry?;
        let src_path = entry.path();
        let dst_path = dst.join(entry.file_name());

        if src_path.is_dir() {
            copy_dir_recursive(&src_path, &dst_path)?;
        } else {
            fs::copy(&src_path, &dst_path)?;
        }
    }

    Ok(())
}
